### ISSUE TITLE
Download file to specific destination

### ISSUE THREAD

**Issue by peterschmidt85** [huggingface/diffusers#2886]
Recently `huggingface_hub. snapshot_download ` added two new arguments: `local_dir` and `local_dir_use_symlinks`. Here's the merged PR: https://github.com/huggingface/huggingface_hub/pull/1360

This is very convenient if you want to use a local folder for storing models instead of the cache folder.

It would be great if `diffusers` would also allow setting these arguments, e.g. with `diffusers.StableUnCLIPPipeline`, `diffusers.StableDiffusionPipeline`,  `diffusers. DiffusionPipeline`, etc.

**Comment by patrickvonplaten** [comment]
cc @Wauplin do you think that's a good idea?

**Comment by Wauplin** [comment]
> This is very convenient if you want to use a local folder for storing models instead of the cache folder.

@peterschmidt85 just to be sure on the use case, when is it preferable to not use the cache folder in your workflow? 

By default `local_dir` will not get rid of the local cache, it just creates symlinks so that you have the weights in a local folder as well. If you really want the actual weights (and not a symlink), that's possible using `local_dir_use_symlinks=False`. But I'm still unsure about the benefit you get from it. In particular, getting rid from the cache system means that the files are re-downloaded each time you re-run the script. It is primarily meant for users that are sure they'll run their script once (example: when building a docker image).

To be clear, I'm not against adding those 2 parameters to diffusion pipelines. I'm just unsure about the use case + I am unsure how it should be documented to make it clear about the limitations.

**Comment by Wauplin** [comment]
In case it helps there is this table [in the docs](https://huggingface.co/docs/huggingface_hub/guides/download#download-files-to-local-folder) to sum it up:

![2023-03-31_09-40](https://user-images.githubusercontent.com/11801849/229055978-ab9a8057-4911-4ae4-93f4-8780194e9cb9.png)

**Comment by peterschmidt85** [comment]
Thank you @Wauplin for the comment.

> when is it preferable not to use the cache folder in your workflow?

The main use-case is not using symlinks

In [dstack](https://github.com/dstackai/dstack), we allow the user to specify which folders to save as artifacts (or to cache). Because `dstack` runs workflows via Docker, artifact paths are mounted via Docker. Both `dstack` and `Docker` do not play nicely with symlinks. Thus, we want to be able to disable them.

Because by default, HuggingFace uses symlinks for `~/cache/huggingface`, we want to be able to save models into another directory so symlinks can be turned off.

**Comment by Wauplin** [comment]
@peterschmidt85 ok, thanks for the context.

If it is for a 1-time setup then maybe it makes sense to use `snapshot_download` directly (with `local_dir_use_symlinks=False`) and load the diffusion pipeline from a folder. Would that be possible on your side?

I'm just afraid that if we implement `diffusers.StableDiffusionPipeline(..., local_dir=..., local_dir_use_symlinks=False)`, users will expect the pipeline to be download the files only once and re-use them each time at each re-run (and we can't blame them for that :smile:). The problem is that it's not how it's working at the moment.

-if this is _really_ a use case widely requested where `snapshot_download` is not an option, we might think of a way of handling this more properly in `huggingface_hub` but it's currently not the case-

**Comment by peterschmidt85** [comment]
@Wauplin The thing is that with the new release of `dstack` we are going to introduce a new feature that will allow the user to cache any local folder.

Example:

```yaml
workflows:
  - name: train
    provider: bash
    commands:
      - pip install -r stable_diffusion/requirements.txt
      - python train.py
    cache:
      - path: ./models/
```

The `train.py` will call `diffusers.StableDiffusionPipeline` and specify the `./models` as a path to load the model to.

The next time the user will run the `train`, `dstack` will automatically mount the cache folder.

This way, the user don't need to call any `snapshot_download`. And this is what I call convenience.

> users will expect the pipeline to be download the files only once and re-use them each time at each re-run (and we can't blame them for that 😄

Not sure I understood what you mean.

**Comment by Wauplin** [comment]
> Not sure I understood what you mean.

Said otherwise, is it ok if each time your users do `pipe = diffusers.StableDiffusionPipeline(..., local_dir=..., local_dir_use_symlinks=False)`, it re-downloads everything? Because that's what will happen if we just pass `local_dir` directly to huggingface_hub.

**Comment by peterschmidt85** [comment]
@Wauplin As a user, when I invoke `pipe = diffusers.StableDiffusionPipeline(..., local_dir=..., local_dir_use_symlinks=False)`, I would assume that if the model is already downloaded in the corresponding folder, it would download it again.

This is exactly how, for example, `torchvision.datasets.MNIST(root=..., download=True)` works. If the data is already downloaded, it won't download it twice.

In my view, this is perfect behavior.

**Comment by peterschmidt85** [comment]
@Wauplin Thanks a lot for the clarification. Yes, indeed.
The whole point is to let the user to opt-out of the default caching mechanism that uses symlinks.

> Out of curiosity, how does it work in torchvision to check if the file has not been updated on their Hub?

Well, `torchvision` uses very simple logic. It re-downloads the entire dataset if any of the files has a different MD5 hash.

> The problem of not using a cache system is that we don't know which version of the file we have on the machine.

Yes, the caching mechanism is super powerful. The only problem with it is that it aggressively uses symlinks.

**Comment by Wauplin** [comment]
> It re-downloads the entire dataset if any of the files has a different MD5 hash.

Gotta be honest, if we investigate further to have a sense of "caching" we would do the same (or at least re-download a file is the hash changed). The problem with computing the file hash each time we load a file is that it takes a lot of time -1.9s for a 500MB file on my laptop for instance-. So if we start to compute all hashes each time we load a diffusion pipeline (that has several files of >1GB), it will be very poor UX as well.

What we might be able to do is to cache tuples `("filepath", "last_modified", "hash")` and trust the OS for the "last_modified". This would be another cache file to store "somewhere". So we have to think more about it before actually starting the implementation.

**Comment by Wauplin** [comment]
In the meantime, I think the quickest workaround for you is to use `snapshot_download` but then your users will have to define which model has to be downloaded :confused:

**Comment by peterschmidt85** [comment]
To re-iterate on it once again, if we want to use Hugging Face in containers, a container-friendly caching mechanism is needed. Consider allowing opt-out from symlinks for running Hugging Face specifically in containers. Unfortunately, snapshot_download doesn't quite do the trick - it doesn't let you reuse cached output.

**Comment by Wauplin** [comment]
> it doesn't let you reuse cached output.

What do you mean by "it doesn't let you reuse cached output?". If users do not care about getting the latest revision once a model is downloaded, they can skip the snapshot_download part. You would have something like:

```py
folder_path = Path(...)
if not folder_path.exists():
    snapshot_download(..., local_dir=folder_path, local_dir_use_symlinks=False)
pipe = StableDiffusionPipeline(...)  # load from pretrained folder
```

It's not perfect but should do the trick, isn't it?

> if we want to use Hugging Face in containers, a container-friendly caching mechanism is needed

We'll keep that in mind. For now we have a lower priority on this as we have other issues to tackle before redefining our cache-system model. I'm sorry but I can't promise we'll deliver a fix for that use case soon. Happy to reconsider if we have high demand and if the `snapshot_download` solution is not enough.

> Consider allowing opt-out from symlinks for running Hugging Face specifically in containers.

It's a bit more complex than that because we still want to keep a reference to know which version of the files are downloaded. This is easy with symlinks but getting rid of them requires a good mechanism. I'll prefer to have more use cases before implementing a "solution to fit them all".
